### PR TITLE
Emulate script fixes

### PR DESCRIPTION
**Pull request by fluidvanadium**

### CODE REVIEW

File: scripts/emulate_app.sh [hunk 1]
```diff
@@ -23,6 +23,7 @@ function check_launch() {
 
 function check_boot() {
     boot_status=$(adb -s emulator-5554 shell getprop sys.boot_completed)
+    echo $boot_status
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: the output from these check functions are `/dev/null` (silenced) so can we delete this line?

File: scripts/emulate_app.sh [hunk 2]
```diff
@@ -177,35 +178,43 @@ if [ ! -d "./android/app" ]; then
     echo "Try './scripts/$(basename $0)' from zingo-mobile root directory." >&2
     exit 1
 fi
-cd android
 
-echo -e "\nInstalling latest build tools, platform tools, and platform..."
-sdkmanager --install 'build-tools;33.0.2' platform-tools
+cd android
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh:181]: can we move this back to the "setup working directory" section? delete space before and add space after

File: scripts/emulate_app.sh [hunk 3]
```diff
@@ -177,35 +178,43 @@ if [ ! -d "./android/app" ]; then
...
 
-echo "Installing system image..."
 avd_name="${avd_skin}-android-${api_level}_${api_target}_${arch}"
 sdk="system-images;android-${api_level};${api_target};${arch}"
-sdkmanager --install "${sdk}"
-sdkmanager --licenses
-
-# Kill all emulators
-../scripts/kill_emulators.sh
+platform="platforms;android-${api_level}"
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh:189]: nice, i like adding platform tools

File: scripts/emulate_app.sh [hunk 4]
```diff
@@ -177,35 +178,43 @@ if [ ! -d "./android/app" ]; then
...
 
 if [[ $create_snapshot == true ]]; then
+    sdkmanager --licenses
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh:192]: i don't think it looks better how the daemon starts running and it prints "All licenses accepted" as the first thing when you run the script. also, all the examples show accepting licenses after the system images. can we put this back to where it was (after installing system images)?

File: scripts/emulate_app.sh [hunk 5]
```diff
@@ -177,35 +178,43 @@ if [ ! -d "./android/app" ]; then
...
-sdkmanager --install 'build-tools;33.0.2' platform-tools
+cd android
+# Kill all emulators
+../scripts/kill_emulators.sh
 
-echo "Installing latest emulator..."
-sdkmanager --install emulator --channel=0
+test_report_dir="app/build/outputs/emulate_app_reports/${abi}"
+rm -rf "${test_report_dir}"
+mkdir -p "${test_report_dir}"
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: why is this moved out of the test and lost its comment? i assume its to do with adding the emulator printouts when creating a snapshot. see below

File: scripts/emulate_app.sh [hunk 6]
```diff
@@ -177,35 +178,43 @@ if [ ! -d "./android/app" ]; then
...
+    sdkmanager --install "${sdk}"
+    sdkmanager --install "${platform}"
+    
     echo -e "\nCreating AVD..."
     echo no | avdmanager create avd --force --name "${avd_name}" --package "${sdk}" --device "${avd_skin}"
 
     echo -e "\n\nWaiting for emulator to launch..."
-    emulator -avd "${avd_name}" -netdelay none -netspeed full -no-window -no-audio -gpu swiftshader_indirect -no-boot-anim \
-        -no-snapshot-load -port 5554 &> /dev/null &
+    emulator -avd "${avd_name}" -netdelay none -netspeed full -no-window -no-audio -no-boot-anim -no-snapshot-load -port 5554 &> "${test_report_dir}/emulator.txt" &
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: i don't like the create snapshot overriding test data or deleting previous test data. can we move the test directory stuff back to where it was and have a similar directory for creating snapshots if we want to capture the printouts?

File: scripts/emulate_app.sh [hunk 7]
```diff
@@ -177,35 +178,43 @@ if [ ! -d "./android/app" ]; then
...
     wait_for $timeout_seconds check_launch
     echo "$(adb devices | grep "emulator-5554" | cut -f1) launch successful"
 
     echo -e "\nWaiting for AVD to boot..."
     wait_for $timeout_seconds check_boot
+    echo -e "\nWaiting for AVD to be online..."
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: `check_boot` and `check_device_online` are two different ways of verifying the same thing (the bootup). i dont think we need this extra line in the console. an improvement would be to combine these two functions into one function where it checks for both at once

File: scripts/emulate_app.sh [hunk 8]
```diff
@@ -232,18 +241,15 @@ else
     ./gradlew assembleDebug -Psplitapk=true
 
     # Create integration test report directory
-    test_report_dir="app/build/outputs/emulate_app_reports/${abi}"
-    rm -rf "${test_report_dir}"
-    mkdir -p "${test_report_dir}"
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh:237]: i think this is where the test report directory should be reset. see above

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]
due to all the installation only being in the `create_snapshot` section, we can't allow the tests to run unless it finds an AVD. can you change the echo send an error message "Error: AVD not found" and then exit the script here:
https://github.com/zingolabs/zingo-mobile/pull/330/files#diff-c978b23d882c1e0ea4a509c0d51c9294acee8ba79ec8a92a0650696ee17b18e7R226

File: scripts/emulate_app.sh (hunk 4 above)
Comment (fluidvanadium) [reply scripts/emulate_app.sh:192]: guess this is a little confusing to me... doesnt it require is to accept the licenses to install the system images?

Comment (Oscar-Pepper) [reply scripts/emulate_app.sh:192]: that would make sense i was just following the "norm" but we can leave it here if you prefer

File: scripts/emulate_app.sh [hunk 9]
```diff
@@ -164,33 +169,21 @@ if [[ $set_abi == false ]]; then
     exit 1
 fi
 
-# Set defaults
-if [[ $set_api_level == false ]]; then
-    api_level=$api_level_default
-fi
-if [[ $set_api_target == false ]]; then
-    api_target=$api_target_default
-fi
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: why have we removed the defaults? this is here to prevent the order of flags being an issue. this introduces the bug where someone puts there `api level` before there `ABI` and it gets reset to default

File: scripts/emulate_app.sh [hunk 10]
```diff
@@ -233,14 +229,12 @@ else
...
     fi
-        
-    echo -e "\nRunning yarn install..."
-    yarn install
-
-    echo -e "\nBuilding APKs..."
-    ./gradlew assembleDebug -Psplitapk=true
 
     # Create integration test report directory
+
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh:235]: random space!

File: scripts/emulate_app.sh (hunk 4 above)
Comment (fluidvanadium) [reply scripts/emulate_app.sh:192]: if it works the other way, no problem.

i could prove or disprove that by deleting sdkmanager and reinstalling from my script, but i dont dare to do that today

File: scripts/emulate_app.sh [hunk 11]
```diff
@@ -249,46 +243,53 @@ else
...
-        fi              
-        if [[ $i -ge 100 ]]; then
-            echo "Error: Failed to install APKs" >&2
-            echo "For more information see 'android/${test_report_dir}/apk_installation.txt'" >&2
-            exit 1
-        fi
-        i=$((i+1))
-        sleep 1
-    done
+    if "${boot_app}"; then
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: can't remember exactly the issue or have the source right now but i have read about issues with boolean in `if` statements with bash, can you add the 'equals true` part like here:
https://github.com/zingolabs/zingo-mobile/pull/330/files/53aef2b69b6c1e1aae75a54e3725edb11a997cd2..5eb69bf6e343c99e65869837ff3b8d6ba816505b#diff-c978b23d882c1e0ea4a509c0d51c9294acee8ba79ec8a92a0650696ee17b18e7R166

File: scripts/emulate_app.sh [hunk 12]
```diff
@@ -249,46 +243,53 @@ else
...
+            if [[ $i -ge 100 ]]; then
+                echo "Error: Failed to install APKs" >&2
+                echo "For more information see 'android/${test_report_dir}/apk_installation.txt'" >&2
+                exit 1
+            fi
+            i=$((i+1))
+            sleep 1
+        done
+
+        # Store emulator info and start logging
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: you probably want to store the emulator info and logs even if you are not booting app?

File: scripts/emulate_app.sh [hunk 13]
```diff
@@ -249,46 +243,53 @@ else
...
         
-    echo -e "\nLaunching App..."
-    nohup adb shell am start -n "org.ZingoLabs.Zingo/org.ZingoLabs.Zingo.MainActivity" -a android.intent.action.MAIN \
-        -c android.intent.category.LAUNCHER > "${test_report_dir}/launch_app.out" &> /dev/null &
+        echo -e "\nLaunching App..."
+        nohup adb shell am start -n "org.ZingoLabs.Zingo/org.ZingoLabs.Zingo.MainActivity" -a android.intent.action.MAIN \
+            -c android.intent.category.LAUNCHER > "${test_report_dir}/launch_app.out" &> /dev/null &
 
-    echo -e "\nTest reports saved: android/${test_report_dir}"        
+        echo -e "\nTest reports saved: android/${test_report_dir}"
```
Comment (Oscar-Pepper) [inline scripts/emulate_app.sh]: as there are things saved to test reports outside of this `boot_app` section, can we put this after we exit this if statement?

also, i just realised it was my mistake but we are using the same directory as integration tests where i copied the script. it would be a nice improvement to have a whole new directory for emulating manually and changing the comment here also:
https://github.com/zingolabs/zingo-mobile/pull/330/files/53aef2b69b6c1e1aae75a54e3725edb11a997cd2..5eb69bf6e343c99e65869837ff3b8d6ba816505b#diff-c978b23d882c1e0ea4a509c0d51c9294acee8ba79ec8a92a0650696ee17b18e7R233

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]
https://github.com/zingolabs/zingo-mobile/pull/330/files/53aef2b69b6c1e1aae75a54e3725edb11a997cd2..5eb69bf6e343c99e65869837ff3b8d6ba816505b#diff-c978b23d882c1e0ea4a509c0d51c9294acee8ba79ec8a92a0650696ee17b18e7R140
this needs updating to explain an AVD must be created before running the emulator

also as mentioned in last review we need to change to an error message and exit if an AVD hasnt been created now as the installation has been moved to `create_snapshot` section

File: .github/workflows/ci.yaml [hunk 14]
```diff
@@ -303,29 +303,9 @@ jobs:
...
-        uses: reactivecircus/android-emulator-runner@v2
-        with:
-          api-level: ${{ matrix.api-level }}
-          arch: ${{ matrix.arch }}
-          force-avd-creation: false
-          emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
-          disable-animations: true
-          script: yarn detox test -c android.emu.debug -n test
-      
+        run: ./scripts/e2e_tests.sh -a ${{ matrix.avd-arch }} -s
```
Comment (Oscar-Pepper) [inline .github/workflows/ci.yaml]: matrix needs to be changed to match integration tests (with `avd-arch` and `abi`  here:
https://github.com/zingolabs/zingo-mobile/pull/330/files/5eb69bf6e343c99e65869837ff3b8d6ba816505b..9490be55d7cd99c4f49e1a7685bec8e22be4617c#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133ddR266

File: scripts/e2e_tests.sh [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+#processing options
+abi_option=" -a x86"
+api_level_option=""
+api_target_option=""
+create_snapshot_option=""
+timeout_seconds_option=""
+while getopts 'e:ba:l:t:sx:h' OPTION; do
+    case "$OPTION" in
+        e)
+            tests_to_run="$OPTARG"
```
Comment (Oscar-Pepper) [inline scripts/e2e_tests.sh]: if `-e` is optional then it needs to be defined if the flag isnt used

File: scripts/e2e_tests.sh [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+        s)  
+            create_snapshot=true
+            create_snapshot_option=" -s"
+            ;;
+        x)
+            timeout_seconds="$OPTARG"
+            ;;
+        h)
+            echo -e "\nE2E test from the command line. Requires Android Studio cmdline-tools."
+            echo -e "\n  -l\t\tSelect tests to run (optional)"
```
Comment (Oscar-Pepper) [inline scripts/e2e_tests.sh]: this flag is `-e` not `-l`

File: scripts/e2e_tests.sh [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+            echo -e "      \t\t  'armeabi-v7a' - default system image: API 30 google_apis_playstore x86"
+            echo -e "\n  -l\t\tSelect API level (optional)"
+            echo -e "      \t\t  Minimum API level: 23"
+            echo -e "\n  -t\t\tSelect API target (optional)"
+            echo -e "      \t\t  See examples on selecting system images below"
+            echo -e "\n  -s\t\tCreate an AVD and snapshot for quick-boot (optional)"
+            echo -e "      \t\t  Does not run app"
+            echo -e "\n  -x\t\tSet timeout in seconds for emulator launch and AVD boot-up (optional)"
+            echo -e "      \t\t  Default: 1800"
+            echo -e "      \t\t  Must be an integer"
```
Comment (Oscar-Pepper) [inline scripts/e2e_tests.sh]: no help for `-b` (reboot)

File: scripts/e2e_tests.sh [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+rm -rf "${test_report_dir}"
+mkdir -p "${test_report_dir}"
+
+avd_name="${avd_skin}-android-${api_level}_${api_target}_${arch}"
+sdk="system-images;android-${api_level};${api_target};${arch}"
+platform="platforms;android-${api_level}"
+
+# "preparing emulator"
+if ! (adb devices | grep emulator-5554) || "${reboot_emulator}"; then
+  emulate_command="./scripts/emulate_app.sh -a ${abi} -l ${api_level} -t ${api_target}${create_snapshot_option} -x ${timeout_seconds} -Y"
```
Comment (Oscar-Pepper) [inline scripts/e2e_tests.sh]: oh ok ive got to this part and ive realised whats going on here...
so can we not have this whole script just be:
1. setup/create avd -> emulate_app.sh -s blah blah blah
2. boot emulator -> emulate_app.sh -Y blah blah blah (using your non-boot-app changes)
3. build detox and prep -> kill all node, start yarn react-native, yarn detox build etc.
4. run e2e tests -> yarn detox test etc.
5. kill emulator when done -> some logic to trigger `kill_emulators.sh`

i think there is lots of unnecessary complications and code that is already handled by the `emulate_app.sh` script, like creating test directories, specifying which ABI etc. 

if anything else is needed for e2e why not add them to the `emulate_app.sh` itself?
maybe we could also simplify the `integraton tests.sh` to use the `emulate_app.sh` further down the line too so there isnt so much duplication of code

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]

**Comment by Oscar-Pepper** [comment]
fixing emulate app script is becoming an e2e PR, might be best to split the PRs?

File: scripts/e2e_tests.sh (hunk 18 above)
Comment (fluidvanadium) [reply scripts/e2e_tests.sh]: youre right, theres a better way to do this

File: scripts/emulate_app.sh (hunk 3 above)
Comment (fluidvanadium) [reply scripts/emulate_app.sh:189]: platform-tools and platforms are different components

**Review by Oscar-Pepper — APPROVED** [review]
great job!
